Add a `nginxprocess` package for working with NGINX processes

ryepup opened the pull request:
### Proposed changes

Add a new `nginxprocess` package to export useful utilities around working with NGINX processes, and update internals to use that package. [NGINXaaS for Azure] has some proprietary code that also wants to use these utilities.

This comes with a significant speed improvement by filtering for NGINX processes earlier and fetching less data from the OS. See individual commits for more details.

[NGINXaaS for Azure]: https://docs.nginx.com/nginx-for-azure/

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [`CONTRIBUTING`](https://github.com/nginx/agent/blob/main/docs/CONTRIBUTING.md) document
- [x] I have run ```make install-tools``` and have attached any dependency changes to this pull request
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works
- [x] If applicable, I have checked that any relevant tests pass after adding my changes
- [x] If applicable, I have updated any relevant documentation ([`README.md`](https://github.com/nginx/agent/blob/main/README.md))
- [x] If applicable, I have tested my cross-platform changes on Ubuntu 22, Redhat 8, SUSE 15 and FreeBSD 13

ryepup commented:
I have hereby read the F5 CLA and agree to its terms

ryepup commented:
I'm not sure why the pipeline is failing; `make unit-test` and `make performance-test` work if I run them locally, and I didn't change the otel collector section.

ryepup commented on internal/watcher/health/nginx_health_watcher_operator.go:
```diff
@@ -39,7 +36,7 @@ func (nhw *NginxHealthWatcher) Health(ctx context.Context, instance *mpi.Instanc
 
 	proc, err := nhw.processOperator.Process(ctx, instance.GetInstanceRuntime().GetProcessId())
-	if errors.Is(err, process.ErrorProcessNotRunning) {
+	if nginxprocess.IsNotRunningErr(err) {
```
I'm matching the existing behavior here, but there's a slim chance that `instance.GetInstanceRuntime().GetProcessId()` is a running process but not an NGINX process. Should we add another branch using `nginxprocess.IsNotNginxErr` to catch that case?

nginx-nickc commented on pkg/nginxprocess/process.go:
```diff
@@ -0,0 +1,142 @@
+}
+
+func convert(ctx context.Context, p *process.Process, o options) (*Process, error) {
+	if ctx.Err() != nil { // fail fast if we've canceled
```
nit,


```
if err := ctx.Err(); err != nil {
	return err
}

```

ryepup replied:
done

nginx-nickc commented on pkg/nginxprocess/process.go:
```diff
@@ -0,0 +1,142 @@
+	var created time.Time
+	if millisSinceEpoch, err := p.CreateTimeWithContext(ctx); err == nil {
+		nanos := millisSinceEpoch * int64(time.Millisecond)
+		created = time.Unix(0, nanos)
```
confused, why not `time. UnixMilli(millisSinceEpoch)`, https://pkg.go.dev/time#UnixMilli

ryepup replied:
> why not ...

because I didn't know `UnixMilli` existed 😄  

done

nginx-nickc commented on pkg/nginxprocess/process.go:
```diff
@@ -0,0 +1,142 @@
+	}
+	for _, p := range processes {
+		pr, cerr := convert(ctx, p, o)
+		if errors.Is(cerr, errNotAnNginxProcess) {
```
use the new func for consistency? `IsNotNginxErr`

ryepup replied:
done

nginx-nickc commented on pkg/nginxprocess/process.go:
```diff
@@ -0,0 +1,142 @@
+	loadStatus bool
+}
+
+// Option customizes how processes are gathered from the OS.
```
Maybe learning for me, the `interface` for `apply` seems extraneous to me. Why not just 

```
type Option func(opts *options)

func WithStatus(v bool) Option {
	return func(opts *options) {
		opts.loadStatus = v
	}
}
```

ryepup replied:
this is the "functional options" implementation we settled on NGINXaaS. It does take more code, but sets a stronger boundary between exported and unexported symbols. This is a similar approach used by [opentelemetry](https://github.com/open-telemetry/opentelemetry-go/blob/c862d2d69cae9f4af2fe8ab244a3b5054d223b55/trace/config.go#L178).

The downside of `type Option func(opts *options)` is that it exposes the unexported `options` type as part of the exported `Option` type. There's not much a caller can do with that information, but it's exposing details we might want to change in the future without having to think about compatibility.

The other (more hypothetical) benefit is flexibility. In the unlikely event this package grows a lot of more features, we could add another option interface (e.g maybe we split into `FindOption` and `ListOption`). `WithStatus` could return a type that implements two interfaces, but that's not possible with `func` types.

If this code were in an `internal` package, I'd probably just use `type Option func(opts *options)`, as all the callers are known. Since this is a public codebase in a package meant to be consumed by other software, I figured I'd err on the side of paranoia.

ryepup commented:
> I'm not sure why the pipeline is failing; `make unit-test` and `make performance-test` work if I run them locally, and I didn't change the otel collector section.

`unit-test` fixed itself, and the `performance-test` was my fault; I added a benchmark that failed if there were no nginx processes. I switched that up so hopefully it'll be happy now.

ryepup commented on pkg/nginxprocess/process.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+// IsWorker returns true if the process is a NGINX worker process.
+func (p *Process) IsWorker() bool { return strings.Contains(p.Cmd, "worker") }
```
this isn't good enough, I can get false positives on the master process using `nginx -c /tmp/workers.conf`

ryepup replied:
got this fixed

ryepup commented on pkg/nginxprocess/process_test.go:
```diff
@@ -0,0 +1,180 @@
+	}
+}
+
+func BenchmarkList(b *testing.B) {
```
This benchmark was useful for development to compare with the existing implementation and verify how expensive additional `gopsutil` calls were. I can see how it might mess up CI due to randomness of the runners.

Which would you prefer:

1. delete it
2. put a `b.Skipf` at the top that devs can comment out if they're working around here
3. put a `if os.Getenv("CI") != "" { b.Skipf(...) }` at the top

dhurley replied:
It they were useful for you during development then maybe we should keep them and put a `b.Skipf` in. Maybe just add a comment as well to the function to say why they are skipped.

ryepup replied:
how does that look?

ryepup commented on pkg/nginxprocess/process_test.go:
```diff
@@ -0,0 +1,180 @@
+		cmd  string
+		want bool
+	}{
+		"nginx_worker": {
```
Will do.

ryepup commented:
Did a rebase, looks like there's some test flake around the otel collector? `panic: close of closed channel`

ryepup commented:
Did another rebase, now getting a different seemingly unrelated test failure with `TestConvertToStructs`. I think that one will fail occasionally because `range` over a `map` is non-deterministic, and the test is expecting it to get keys in a certain order.

ryepup commented:
opened #968 to fix the test thats broken my latest pipeline
